### ISSUE TITLE
Expanding multiple levels (support $levels option for $expand)

### ISSUE THREAD

**Issue by Dimsday** [object/Simple.OData.Client#85]
Please add support for $levels option for expand more the one related entites (e.g. for hierarchial data).  
Url like:
http://host/odata/Entity?$expand=Children($levels=2)
or
http://host/odata/Entity?$expand=Children($levels=max)

http://docs.oasis-open.org/odata/odata/v4.0/errata02/os/complete/part2-url-conventions/odata-v4.0-errata02-os-part2-url-conventions-complete.html#_Toc406398162

**Comment by object** [comment]
Thank you for the suggestion. I will look into it.

**Comment by Dimsday** [comment]
There is here a class (suddenly it is useful):
```
Microsoft.OData.Core.UriParser.Semantic.LevelsClause
```

**Comment by object** [comment]
I am trying to figure out how this can fit in current S.O.D. syntax, e.g. statements like this:

For<People>().Expand(x => x.Friends)

Expand should probably get a second (optional) argument for levels, but then there may be multiple expands with different levels.

**Comment by Dimsday** [comment]
```
For<People>().Expand(x => x.Friends, level)
```
or
```
For<People>().Expand(x => x.Friends).ExpandLevel(level)
```
"level" - may be stuct like
```
Microsoft.OData.Core.UriParser.Semantic.LevelsClause
```

**Comment by object** [comment]
If there is a new clause, like ExpandLevel, perhaps there will be needed an interface change to allow use of ExpandLevel only after Expand. Or just extend the Expand clause with an optional argument.

Decisions, decisions...

**Comment by Dimsday** [comment]
I think you can not restrict the use of ExpandLevel. Just ignore it, unless Expand.
But the version with an additional optional parameter looks attractive:-)

**Comment by object** [comment]
FYI: support for multiple levels is coming in the version 4.5 which will be out sometime next week.

**Comment by object** [comment]
Implemented in 4.5.

**Comment by Allann** [comment]
Thank you for adding this.  Whilst I like this feature, I can't figure out how it will work in the following instance:
            var environments = await _client
                .For<ServerEnvironment>()
                .Filter(se => se.Module.Code == moduleCode)
                .Expand(x => new
                    {
                        x.Site,
                        x.EnvironmentType,
                        x.AssociatedEnvironments
                    })
                .FindEntriesAsync();

Because I need to expand numerous children on the initial ServerEnvironment, and then need to expand the children of the AssociatedEnvironments

As an example you can use the following URL:
http://jbssecurity.azurewebsites.net/ServerEnvironments?$filter=Module/Code%20eq%20%27A0002%27%20and%20EnvironmentType/Code%20eq%20%27DEV%27&$expand=Site,EnvironmentType,AssociatedEnvironments($expand=Module)

Basically I can't figure out how to get the final expand on the AssociatedEnvironments.

**Comment by object** [comment]
When expanding nested collections, you can use LINQ Select clause, like in this example:

var product = (await _client
                .For&lt;Product&gt;()
                .OrderBy(x =&gt; x.ProductID)
                .Expand(x =&gt; x.Category.Products.Select(y =&gt; y.Category))
                .FindEntriesAsync()).Last();
